main/src/library/List.flix

```diff
@@ -0,0 +1,579 @@
+
+    ///
+    /// Helper function for `mapWithIndex`.
+    /// Returns the result of applying `f` to every element in `xs` along with that element's index + `i`.
+    /// That is, the result is of the form: `f(x1, i) :: f(x2, i+1) :: ...`.
+    /// mapWithIndex(f, xs) = mapWithIndexHelper(f, xs, 0).
+    ///
+    def mapWithIndexHelper[a,b](f: (a, Int32) -> b, xs: List[a], i: Int32): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => f(x, i) :: mapWithIndexHelper(f, rs, i+1)
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs` and concatenating the results.
+    ///
+    def flatMap[a,b](f: a -> List[b], xs: List[a]): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => append(f(x), flatMap(f, rs))
+    }
+
+    ///
+    /// Returns the reverse of `xs`.
+    ///
+    def reverse[a](xs: List[a]): List[a] = reverseHelper(xs, Nil)
+
+    ///
+    /// Helper function for `reverse`.
+    /// Returns `acc` appended to the reverse of `xs`.
+    /// reverse(xs) = reverseHelper(xs, Nil).
+    ///
+    def reverseHelper[a](xs: List[a], acc: List[a]): List[a] = match xs with {
+        case Nil => acc
+        case x :: rs => reverseHelper(rs, x :: acc)
+    }
+
+    ///
+    /// Returns `xs` with its elements rotated `n` positions to the left.
+    /// That is, returns a new list where the first `n mod length(xs)` elements in `xs`
+    /// are the last `n mod length(xs)` elements of the new list.
+    ///
+    def rotateLeft[a](n: Int32, xs: List[a]): List[a] =
+        let len = length(xs);
+        if (len == 0) xs else
+            let rem = n % len;
+            let rotate = if (rem < 0) rem + len else rem;
+            append(drop(rotate, xs), take(rotate, xs))
+
+    ///
+    /// Returns `xs` with its elements rotated `n` positions to the right.
+    /// That is, returns a new list where the last `n mod length(xs)` elements in `xs`
+    /// are the first `n mod length(xs)` elements of the new list.
+    ///
+    def rotateRight[a](n: Int32, xs: List[a]): List[a] = rotateLeft(-n, xs)
+
+    ///
+    /// Returns `xs` with the element at index `i` replaced by `x`.
+    /// Returns `xs` if `i < 0` or `i > length(xs)-1`.
+    ///
+    def replace[a](i: Int32, x: a, xs: List[a]): List[a] = match (i, xs) with {
+        case (_, Nil) => xs
+        case (0, r :: rs) => x :: rs
+        case (p, r :: rs) => r :: replace(p-1, x, rs)
+    }
+
+    ///
+    /// Returns `ys` with the `n` elements starting at index `i` replaced with the elements of `xs`.
+    /// If any of the indices `i, i+1, i+2, ... , i+n-1` are out of range in `ys` then no patching is done at these indices.
+    /// If `xs` becomes depleted then no further patching is done.
+    /// If patching occurs at index `i+j` in `ys`, then the element at index `j` in `xs` is used.
+    ///
+    def patch[a](i: Int32, n: Int32, xs: List[a], ys: List[a]): List[a] = patchHelper(i, n, drop(-i, xs), ys, 0)
```

Comment:
`patch` is now valid for any set of inputs.  All or part of `xs` is simply mapped/patched into a range of `ys`.  Anywhere the range is invalid, no patching occurs.